### PR TITLE
feat: add datatables component

### PR DESCRIPTION
**Pull request by jonathanzhang53**
## Description
Implemented datatables component as shown in #170.

## Related issues and discussion
#170 

## GIF
asdwfger

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

**Comment by jonathanzhang53** [comment]
@abhaychawla I am unsure what to do next right now. I am currently working on this component while using financial-activity-mappings in accounting as an example. 

Sorry for misunderstanding the task, but I am confused on what changes I need to make in order for the table to be able to 'read' data on datatables. I have currently made the UI for the table, but am I correct when I say that manage-data-tables.component.ts is still not complete? I have tried making the backend of the table but I get some errors when refactoring code from financial-activity-mappings. 

Any tips for me on how to approach this? I am new to developing APIs.

**Comment by jonathanzhang53** [comment]
@abhaychawla I have emulated the code from financial-activity-mappings, and I have added the necessary functions for the table to be initialized, but I don't know whether it is what you meant when you said 'making requests through an API you'll write its code within the system service and use a resolver to fetch the required data' on my GCI task. Please look over when you have time and let me know what I need to change/add 😄

**Comment by abhaychawla** [comment]
Jonathan, when you send a request to `/datatables`, the data for all existing data tables will be returned.
API: https://demo.openmf.org/fineract-provider/api/v1/datatables?tenantIdentifier=default&pretty=true
As you have written the required code in the service, now this data needs to be shown in the table. See how resolvers are used to fetch the data in angular and make the data visible in the table by using the required properties of the object that was returned when you sent the API request. To see which properties of the object need to be used you'll have to look for the implementation of data tables in the community-app.

**Comment by jonathanzhang53** [comment]
@abhaychawla How will I get data to show in the table when there is no way to create a datatable ?

**Comment by jonathanzhang53** [comment]
I have added the {{ }} fields to the component but they will have no values because the create component has not been made.

**Comment by abhaychawla** [comment]
We are currently working with the demo server. It already contains some data, as you can see here: https://demo.openmf.org/#/datatables

**Comment by jonathanzhang53** [comment]
@abhaychawla So I need to inject the data in somehow? I'm not clear on how to do that 😝

**Comment by abhaychawla** [comment]
That's what I am saying do some research, read angular's documentation on how to get the data using resolvers and have a look at how its implemented here. Then try to do that for this component. The property names for the object that you need to access which are visible in the table also need to be looked into the community-app. The current properties are incorrect. The heading of column and object property are not necessarily always same.

**Comment by jonathanzhang53** [comment]
@abhaychawla Okay I will try my best :) See you tomorrow.

**Comment by abhaychawla** [comment]
Take your time, and if you are not willing to work on the API functionality, just submit the code for the user interface as its mentioned in the task. Remove the extra code in that case.

**Comment by jonathanzhang53** [comment]
@abhaychawla I'll do some research on the API functionality.
Am I correct to fix these areas in the code?

html component:

      <ng-container matColumnDef="dataTableName">
        <th mat-header-cell *matHeaderCellDef mat-sort-header> Data Table Name </th>
        <td mat-cell *matCellDef="let dataTable"> {{ dataTable.name }} </td>
      </ng-container>

      <ng-container matColumnDef="associatedWith">
        <th mat-header-cell *matHeaderCellDef mat-sort-header> Associated With </th>
        <td mat-cell *matCellDef="let dataTable"> {{ dataTable.associatedWith }} </td>
      </ng-container>

ts component:

```
  /**
   * Initializes the data source, paginator and sorter for manage data tables table.
   */
  setDataTables() {
    this.dataSource = new MatTableDataSource(this.dataTableData);
    this.dataSource.paginator = this.paginator;
    this.dataSource.sortingDataAccessor = (dataTable: any, property: any) => {
      switch (property) {
        case 'dataTableName': return dataTable.name;
        case 'associatedWith': return dataTable.associatedWith;
        default: return dataTable[property];
      }
    };
    this.dataSource.sort = this.sort;
  }
```

I just want to make sure I'm approaching the issue correctly.

**Comment by jonathanzhang53** [comment]
@abhaychawla I GOT IT TO WORK. SEE UPDATED GIF !!!

Are you still awake?

**Comment by jonathanzhang53** [comment]
@abhaychawla I have a question, where did you get the api link that you posted: https://demo.openmf.org/fineract-provider/api/v1/datatables?tenantIdentifier=default&pretty=true

File: src/app/system/manage-data-tables/manage-data-tables.component.html [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    </mat-form-field>
+  </div>
+
+  <div class="mat-elevation-z8">
+
+    <table mat-table [dataSource]="dataSource" matSort>
+
+      <ng-container matColumnDef="dataTableName">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header> Data Table Name </th>
+        <td mat-cell *matCellDef="let dataTable"> {{ dataTable.applicationTableName }} </td>
```
Comment (abhaychawla) [inline src/app/system/manage-data-tables/manage-data-tables.component.html]: You have swapped the properties. This should be registeredTableName.
See here: https://github.com/openMF/web-app/issues/170

File: src/app/system/manage-data-tables/manage-data-tables.component.html [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      </ng-container>
+
+      <ng-container matColumnDef="associatedWith">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header> Associated With </th>
+        <td mat-cell *matCellDef="let dataTable"> {{ dataTable.registeredTableName }} </td>
```
Comment (abhaychawla) [inline src/app/system/manage-data-tables/manage-data-tables.component.html]: You have swapped the properties. This should be applicationTableName.
See here: https://github.com/openMF/web-app/issues/170

File: src/app/system/manage-data-tables/manage-data-tables.component.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    this.setDataTables();
+  }
+
+  /**
+   * Initializes the data source, paginator and sorter for manage data tables table.
+   */
+  setDataTables() {
+    this.dataSource = new MatTableDataSource(this.dataTableData);
+    this.dataSource.paginator = this.paginator;
+    this.dataSource.sortingDataAccessor = (dataTable: any, property: any) => {
```
Comment (abhaychawla) [inline src/app/system/manage-data-tables/manage-data-tables.component.ts]: We usually use this in the case of nested objects only. Sorting can be performed in an easier way without the need for a sorting data accessor if the object is not nested by just specifying the name of the property in the column definition. See https://github.com/openMF/web-app/blob/master/src/app/accounting/closing-entries html and ts file as an example and make the necessary changes.

File: src/app/system/system-routing.module.ts [hunk 4]
```diff
@@ -48,6 +50,14 @@ const routes: Routes = [
           component: ExternalServicesComponent,
           data: { title:  extract('External Services'), breadcrumb: 'External Services' },
       },
+      {
+          path: 'manage-data-tables',
```
Comment (abhaychawla) [inline src/app/system/system-routing.module.ts]: Change to `data-tables`.

File: src/app/system/manage-data-tables/manage-data-tables.component.html [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      <tr mat-header-row *matHeaderRowDef="displayedColumns"></tr>
+      <tr mat-row *matRowDef="let row; columns: displayedColumns;" class="select-row"></tr>
+
+    </table>
+
+    <mat-paginator [pageSizeOptions]="[10, 25, 50, 100]" showFirstLastButtons></mat-paginator>
+
+  </div>
+
+</div>
```
Comment (abhaychawla) [inline src/app/system/manage-data-tables/manage-data-tables.component.html]: Ensure you end the file with a new line.

**Comment by jonathanzhang53** [comment]
@abhaychawla made the changes. just need to update the gif quickly. :)

**Comment by jonathanzhang53** [comment]
@abhaychawla for some reason, the sorting broke when I tried to emulate closing entries. LOoking into it right now.

**Comment by jonathanzhang53** [comment]
@abhaychawla I'm really worried that I won't be able to make these changes in time because I am working on this at school, and I have to use VPN because school blocks me from running localhost api requests. Will you be awake for another hour?

**Comment by abhaychawla** [comment]
Yes, I am available. I had an exam today so was unable to get back to you yesterday night.

**Comment by jonathanzhang53** [comment]
@abhaychawla that's alright. I have looked over closing-entries and my sort doesn't work for some reason. The data is displaying but not sorting when you click the column titles and I have defined them in html.

**Comment by abhaychawla** [comment]
As I mentioned above, you need to make changes to the column definition. Have a look at the property names and the column definition. They should be the same for sorting to work without the need of an accessor to map the values.

**Comment by jonathanzhang53** [comment]
@abhaychawla I think I see what you're talking about. I am a bit frustrated because at school the loading for the datatables data takes so long.

**Comment by jonathanzhang53** [comment]
@abhaychawla fixed.

**Comment by jonathanzhang53** [comment]
@abhaychawla thanks for the help. I have posed a question on issue #195 about an API request. Not sure why I cannot fetch product mix data :)

**Pull request merged**